packages/react/src/components/TimePicker/TimePicker.tsx

```diff
@@ -0,0 +1,134 @@
+import * as React from 'react';
+import { ComboBox, IComboBox, IComboBoxOption } from '@fluentui/react/lib/ComboBox';
+import { TimeConstants, addMinutes, formatTimeString, ceilMinute } from '@fluentui/date-time-utilities';
+import { ITimePickerProps, TimeRange } from './TimePicker.types';
+
+// Valid KeyChars for user input
+const KEYCHAR_0 = 48;
+const KEYCHAR_COLON = 58;
+const KEYCHAR_SPACE = 32;
+const KEYCHAR_A = 65;
+const KEYCHAR_M = 77;
+const KEYCHAR_P = 80;
+
+export const TimePicker = ({
+  label,
+  increments = 30,
+  showSeconds = false,
+  allowFreeform = true,
+  useHour12 = false,
+  onFormatDate,
+  timeRange = { start: -1, end: -1 },
+  ...rest
+}: ITimePickerProps) => {
+  const [selectedKey, setSelectedKey] = React.useState<string | number | undefined>(0);
+  const [userText, setUserText] = React.useState<string>('');
+  const [errorMessage, setErrorMessage] = React.useState<string>('');
+
+  const defaultTime = generateDefaultTime(increments, timeRange);
+  const optionsCount = getDropdownOptionsCount(increments, timeRange);
+  const timePickerOptions: IComboBoxOption[] = Array(optionsCount)
+    .fill(0)
+    .map((_, index) => {
+      const option = addMinutes(defaultTime, increments * index);
+      option.setSeconds(0);
+      return {
+        key: index,
+        text: onFormatDate ? onFormatDate(option) : `${formatTimeString(option, showSeconds, useHour12)}`,
+      };
+    });
+
+  const onChange = React.useCallback(
+    (event: React.FormEvent<IComboBox>, option?: IComboBoxOption, index?: number, value?: string): void => {
+      let key = option?.key;
+      let updatedUserText = '';
+      if (value) {
+        if (allowFreeform && !option && !onFormatDate) {
+          // Validate only if user did not add onFormatDate
+          validateUserInput(value, showSeconds, useHour12);
+        }
+        updatedUserText = value;
+      } else if (option) {
+        updatedUserText = option.text;
+        setErrorMessage('');
+      }
+      setUserText(updatedUserText);
+      setSelectedKey(key);
+    },
+    [allowFreeform],
+  );
+
+  const evaluatePressedKey = (event: React.KeyboardEvent<IComboBox>) => {
+    if (
+      !onFormatDate &&
+      // Only permit input of digits, space, colon, A/P/M characters
+      !(
+        (event.charCode >= KEYCHAR_0 && event.charCode <= KEYCHAR_COLON) ||
+        event.charCode === KEYCHAR_SPACE ||
+        event.charCode === KEYCHAR_A ||
+        event.charCode === KEYCHAR_M ||
+        event.charCode === KEYCHAR_P
+      )
+    ) {
+      event.preventDefault();
+    }
+  };
+
+  const validateUserInput = (userInput: string, showSeconds: boolean, useHour12: boolean) => {
```

Comment:
This seems like it might be something that would be worth allowing a consumer to override. If you're letting them format the date, might as well let them do their own custom validation as well (it looks like currently we just skip over validation if we're doing custom formatting?)

Reply:
Resolved by adding `onValidationUserInput` optional prop for users to override validation behavior, but `onValidationUserInput` only runs when user also provides optional prop `onFormatDate`.